torchdata/nodes/base_node.py

```diff
@@ -4,33 +4,88 @@
 # This source code is licensed under the BSD-style license found in the
 # LICENSE file in the root directory of this source tree.
 
-from typing import Iterable, Iterator, TypeVar
+import logging
+from typing import Any, Dict, Iterable, Iterator, Optional, TypeVar
+
+logger = logging.getLogger(__name__)
 
 
 T = TypeVar("T", covariant=True)
 
 
+class BaseNodeIterator(Iterator[T]):
+    def state_dict(self) -> Dict[str, Any]:
+        raise NotImplementedError()
+
+    def started(self) -> bool:
+        raise NotImplementedError()
+
+    def finished(self) -> bool:
+        raise NotImplementedError()
+
+
 class BaseNode(Iterable[T]):
-    def iterator(self) -> Iterator[T]:
+    __it: Optional[BaseNodeIterator[T]] = None  # holds pointer to last iter() requested
+    __initial_state: Optional[Dict[str, Any]] = None
+
+    def iterator(self, initial_state: Optional[dict]) -> Iterator[T]:
         """Override this method to implement the iterator.
         Iterators are expected to raise StopIteration to signal
         end of iteration, so they can be used in for loops.
         Generators just need to return, as usual.
+
+        initial_state will be passed if `load_state_dict(initial_state)` was called
+        on this node before the __iter__ is requested, otherwise None will be passed
         """
         raise NotImplementedError()
 
-    def __iter__(self) -> "_EagerIter[T]":
-        return _EagerIter(self)
+    def get_state(self) -> Dict[str, Any]:
+        """Return a dictionary that can be passed to iterator(...) which
+        can be used to initialize iterator at a certain state.
+        """
+        raise NotImplementedError()
+
+    def __iter__(self) -> BaseNodeIterator[T]:
+        if self.__it is not None and self.__it.started():
+            # Only create a new iter if the last requested one did not start/finish
```

Comment:
Confirming, we return a new iterator when the previous one finished?

Reply:
yes, or even if it just started, eg imagine this scenario:
```
ds = IterableWrapper(range(10))
it = iter(ds)
for _ in range(5):
  next(it)

list(ds) # Should result in [0, ..., 10]
```

Reply:
Thanks. Thats basically saying that unless you save the state, or load a new 'initial state' we start over everytime you request a new iter (which is the desired behavior).